cli_tools/common/utils/storage/storage_client.go

```diff
@@ -129,6 +130,43 @@ func (sc *Client) DeleteGcsPath(gcsPath string) error {
 	return nil
 }
 
+// DeleteObject deletes the object with the path `gcsPath`. Deletion requires a single
+// exact match on the path. If the object isn't found, an error is returned.
+// If multiple objects are found with `gcsPath` as a prefix, no deletion occurs,
+// and an error is returned.
+func (sc *Client) DeleteObject(gcsPath string) error {
+	bucketName, objectPath, err := SplitGCSPath(gcsPath)
+	if err != nil {
+		return err
+	}
+	it := sc.GetObjects(bucketName, objectPath)
+
+	firstObject, err := it.Next()
+	if err == iterator.Done {
+		return daisy.Errf("Error deleting `%v`: Object not found", gcsPath)
+	}
+	if err != nil {
+		return daisy.Errf("Error deleting `%v`: Failed querying GCS: `%v`", gcsPath, err)
+	}
+	if firstObject == nil {
+		return daisy.Errf("Error deleting `%v`: Object not found", gcsPath)
+	}
+	_, err = it.Next()
+	if err == nil {
+		return daisy.Errf("Error deleting `%v`: Multiple objects with prefix", gcsPath)
+	}
+	if err != iterator.Done {
+		return daisy.Errf("Error deleting `%v`: Failed querying GCS: `%v`", gcsPath, err)
+	}
+	if firstObject.Name != objectPath {
+		return daisy.Errf("Error deleting `%v`: Object not found", gcsPath)
+	}
+	if err := sc.GetObject(bucketName, firstObject.Name).Delete(); err != nil {
```

Comment:
GetObject().Delete() should delete only one object. I don't believe it works on prefixes, so all the iteration and ensuring there's only one object is unnecessary IMO. 

One of the reason why DeleteObjects() had to iterate is because of no prefix deletion.

Reply:
If `-source_file` is an object reference, but also a prefix for other objects, then `GetObject().Delete()` would remove the directly-referenced object, while leaving the partially-matched objects. There are some examples of this in `scrach_ownership_test`; see:
  - Test_DontDeleteSourceFile_WhenAnotherFileHasSamePrefix
  - Test_DontDeleteSourceFile_WhenDirectory when `makeDirObject` is true

I suppose there's an argument that it's alright to delete in both of these cases (it's only one file). I decided against that since neither of those cases occur from  `gcloud compute images import`.

Regarding the new library method `DeleteObject`, I'd argue that this implementation is the safest choice. Consider when a dev that doesn't read the docs, and passes a directory object ('path/' referes to an empty object), assuming that the entire directory will get removed. Without the safety of the current method, the unsuspective dev would think the directory had been cleaned up, but in reality all of the children objects are still present.

Reply:
DeleteObject() should delete one object, irrespective if another object has the same file prefix. It should not throw an error, as that makes it impossible to delete a single file in all cases with this API. We already have DeleteGcsPath() for the prefix deletion. Func name and docs should be enough to clarify that DeleteObject() will delete a single file only.

Reply:
I think Zoran's point is that, if you found a exact match, then you don't need to care about any prefix match.

Reply:
Talked with Zoran offline; the current version has the following modifications:
1. storage_client.DeleteObject does an exact lookup on the object path
2. Scratch cleanup only occurs when the client is gcloud.